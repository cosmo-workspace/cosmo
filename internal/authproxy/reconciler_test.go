package authproxy

import (
	"context"
	"time"

	wsv1alpha1 "github.com/cosmo-workspace/cosmo/api/workspace/v1alpha1"
	"github.com/cosmo-workspace/cosmo/internal/authproxy/proxy"
	"github.com/cosmo-workspace/cosmo/pkg/auth"
	"github.com/cosmo-workspace/cosmo/pkg/clog"
	"github.com/cosmo-workspace/cosmo/pkg/kosmo"
	"github.com/cosmo-workspace/cosmo/pkg/kubeutil"
	"k8s.io/utils/pointer"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/healthz"
	logf "sigs.k8s.io/controller-runtime/pkg/log"
	"sigs.k8s.io/controller-runtime/pkg/log/zap"

	. "github.com/cosmo-workspace/cosmo/pkg/snap"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

var _ = Describe("auth-proxy controller", func() {

	var (
		ctx          context.Context
		cancel       context.CancelFunc
		clientMock   kubeutil.ClientMock
		proxyManager *proxy.Manager
	)

	startManager := func(username, workspaceName string) {
		By("---- Manager setting start ----")
		logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))

		mgr, err := ctrl.NewManager(cfg, ctrl.Options{
			Scheme:             scheme,
			MetricsBindAddress: "0",
			LeaderElection:     false,
			Namespace:          "cosmo-user-" + username,
			//		CertDir:            testEnv.WebhookInstallOptions.LocalServingCertDir,
			Port: testEnv.WebhookInstallOptions.LocalServingPort,
		})
		Expect(err).NotTo(HaveOccurred())

		clientMock = kubeutil.NewClientMock(mgr.GetClient())
		klient := kosmo.NewClient(&clientMock)

		authorizer := auth.NewPasswordSecretAuthorizer(mgr.GetClient())
		proxyManager, err = (&proxy.Manager{
			Log:                      clog.NewLogger(ctrl.Log.WithName("proxy-manager")),
			ProxyBackendScheme:       "http",
			ProxyGracefulShutdownDur: time.Second * time.Duration(10),
			ProxyStartupCheckTimeout: time.Second * time.Duration(10),
			Insecure:                 true,
			TLSCertPath:              "",
			TLSPrivateKeyPath:        "",
			User:                     username,
			MaxAgeSeconds:            60 * 720,
			Authorizer:               authorizer,
		}).Initialize()
		Expect(err).NotTo(HaveOccurred())

		reconciler := NetworkRuleReconciler{
			Client:        klient,
			Recorder:      mgr.GetEventRecorderFor("cosmo-auth-proxy"),
			Scheme:        mgr.GetScheme(),
			ProxyManager:  proxyManager,
			WorkspaceName: workspaceName,
		}
		err = reconciler.SetupWithManager(mgr)
		Expect(err).NotTo(HaveOccurred())

		err = mgr.AddHealthzCheck("healthz", healthz.Ping)
		Expect(err).NotTo(HaveOccurred())
		err = mgr.AddReadyzCheck("readyz", healthz.Ping)
		Expect(err).NotTo(HaveOccurred())

		ctx, cancel = context.WithCancel(context.Background())
		By("---- Manager setting complete ----")

		go func() {
			By("----Start Manager----")
			err := mgr.Start(ctx)
			Expect(err).NotTo(HaveOccurred())
		}()
	}

	stopManager := func() {
		By("----Stop Manager in porgress---")
		cancel()
		time.Sleep(100 * time.Millisecond)
	}

	workspaceSnap := func(ws *wsv1alpha1.Workspace) struct{ Name, Namespace, Spec, Status interface{} } {
		if ws == nil {
			return struct{ Name, Namespace, Spec, Status interface{} }{}
		}
		ws = ws.DeepCopy()
		for i, nw := range ws.Spec.Network {
			if nw.TargetPortNumber != nil && *nw.TargetPortNumber != int32(nw.PortNumber) {
				ws.Spec.Network[i].TargetPortNumber = pointer.Int32(99999)
			}
		}
		return struct{ Name, Namespace, Spec, Status interface{} }{
			Name:      ws.Name,
			Namespace: ws.Namespace,
			Spec:      ws.Spec,
			Status:    ws.Status,
		}
	}

	proxySnap := func(proxies []proxy.LocalPortProxyInfo) []proxy.LocalPortProxyInfo {
		snapProxies := make([]proxy.LocalPortProxyInfo, 0, len(proxies))
		for _, p := range proxies {
			if p.LocalPort != 0 {
				p.LocalPort = 99999
			}
			snapProxies = append(snapProxies, p)
		}
		return snapProxies
	}

	//==================================================================================

	AfterEach(func() {
		clientMock.Clear()
		testUtil.DeleteWorkspaceAll()
		testUtil.DeleteCosmoUserAll()
		testUtil.DeleteTemplateAll()
	})

	//==================================================================================

	DescribeTable("✅ success in normal context:",

		func(initialFunc func(), modifyFuncs ...func()) {
			testUtil.CreateTemplate(wsv1alpha1.TemplateTypeWorkspace, "template1")
			testUtil.CreateLoginUser("test-user", "", "", "password")
			initialFunc()
			By("---------------test start----------------")
			defer stopManager()
			startManager("test-user", "test-workspace")
			time.Sleep(time.Second * 3)

			for _, modifyFunc := range modifyFuncs {
				modifyFunc()
				time.Sleep(time.Second * 3)

				wsv1Workspace, _ := kosmoClient.GetWorkspaceByUserName(context.Background(), "test-workspace", "test-user")
				Ω(workspaceSnap(wsv1Workspace)).To(MatchSnapShot())

				proxies := proxyManager.GetRunningProxies()
				Ω(proxySnap(proxies)).To(MatchSnapShot())
			}
			By("---------------test end---------------")
		},

		Entry("01 when starting a reconcile after creating network rules",
			func() {
				testUtil.CreateWorkspace("test-user", "test-workspace", "template1", nil)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw1", 1111, "gp1", "/", false)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw2", 2222, "gp2", "/", true)
			},
			func() {},
		),

		Entry("02 when starting a reconcile before creating network rules",
			func() {},
			func() {
				testUtil.CreateWorkspace("test-user", "test-workspace", "template1", nil)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw1", 1111, "gp1", "/", false)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw2", 2222, "gp2", "/", true)
			},
		),

		Entry("03 when changing the port Number",
			func() {
				testUtil.CreateWorkspace("test-user", "test-workspace", "template1", nil)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw1", 1111, "gp1", "/", false)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw2", 2222, "gp2", "/", true)
			},
			func() {
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw1", 3333, "gp1", "/", false)
			},
			func() {
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw2", 4444, "gp2", "/", true)
			},
		),

		Entry("04 when changing the path ",
			func() {
				testUtil.CreateWorkspace("test-user", "test-workspace", "template1", nil)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw1", 1111, "gp1", "/", false)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw2", 2222, "gp2", "/", true)
			},
			func() {
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw1", 1111, "gp1", "/abc", false)
			},
			func() {
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw2", 2222, "gp2", "/cdf", true)
			},
		),

		Entry("05 when changing the public flag",
			func() {
				testUtil.CreateWorkspace("test-user", "test-workspace", "template1", nil)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw1", 1111, "gp1", "/", false)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw2", 2222, "gp2", "/", true)
			},
			func() {
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw1", 1111, "gp1", "/", true)
			},
			func() {
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw2", 2222, "gp2", "/", false)
			},
		),

		Entry("06 wheh deleting the network rule",
			func() {
				testUtil.CreateWorkspace("test-user", "test-workspace", "template1", nil)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw1", 1111, "gp1", "/", false)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw2", 2222, "gp2", "/", true)
			},
			func() {
				testUtil.DeleteNetworkRule("test-user", "test-workspace", "nw2")
			},
			func() {
				testUtil.DeleteNetworkRule("test-user", "test-workspace", "nw1")
			},
		),

		Entry("07 when deleting the workspace",
			func() {},
			func() {
				testUtil.CreateWorkspace("test-user", "test-workspace", "template1", nil)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw1", 1111, "gp1", "/", false)
				testUtil.UpsertNetworkRule("test-user", "test-workspace", "nw2", 2222, "gp2", "/", true)
			},
			func() {
				testUtil.DeleteWorkspaceAll()
			},
		),
	)

})
