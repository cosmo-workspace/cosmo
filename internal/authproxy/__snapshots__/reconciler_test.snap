['auth-proxy controller ✅ success in normal context: 01 when starting a reconcile after creating network rules 1']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1,
    \"network\": [
      {
        \"name\": \"nw1\",
        \"portNumber\": 1111,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 99999,
        \"host\": \"gp1-test-workspace-test-user.domain\",
        \"group\": \"gp1\",
        \"public\": false
      },
      {
        \"name\": \"nw2\",
        \"portNumber\": 2222,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 2222,
        \"host\": \"gp2-test-workspace-test-user.domain\",
        \"group\": \"gp2\",
        \"public\": true
      }
    ]
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 01 when starting a reconcile after creating network rules 2']
SnapShot = """
[
  {
    \"Name\": \"proxy1111\",
    \"TargetPort\": 1111,
    \"LocalPort\": 99999
  }
]
"""

['auth-proxy controller ✅ success in normal context: 02 when starting a reconcile before creating network rules 1']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1,
    \"network\": [
      {
        \"name\": \"nw1\",
        \"portNumber\": 1111,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 99999,
        \"host\": \"gp1-test-workspace-test-user.domain\",
        \"group\": \"gp1\",
        \"public\": false
      },
      {
        \"name\": \"nw2\",
        \"portNumber\": 2222,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 2222,
        \"host\": \"gp2-test-workspace-test-user.domain\",
        \"group\": \"gp2\",
        \"public\": true
      }
    ]
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 02 when starting a reconcile before creating network rules 2']
SnapShot = """
[
  {
    \"Name\": \"proxy1111\",
    \"TargetPort\": 1111,
    \"LocalPort\": 99999
  }
]
"""

['auth-proxy controller ✅ success in normal context: 03 when changing the port Number 1']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1,
    \"network\": [
      {
        \"name\": \"nw1\",
        \"portNumber\": 3333,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 99999,
        \"host\": \"gp1-test-workspace-test-user.domain\",
        \"group\": \"gp1\",
        \"public\": false
      },
      {
        \"name\": \"nw2\",
        \"portNumber\": 2222,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 2222,
        \"host\": \"gp2-test-workspace-test-user.domain\",
        \"group\": \"gp2\",
        \"public\": true
      }
    ]
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 03 when changing the port Number 2']
SnapShot = """
[
  {
    \"Name\": \"proxy3333\",
    \"TargetPort\": 3333,
    \"LocalPort\": 99999
  }
]
"""

['auth-proxy controller ✅ success in normal context: 03 when changing the port Number 3']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1,
    \"network\": [
      {
        \"name\": \"nw1\",
        \"portNumber\": 3333,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 99999,
        \"host\": \"gp1-test-workspace-test-user.domain\",
        \"group\": \"gp1\",
        \"public\": false
      },
      {
        \"name\": \"nw2\",
        \"portNumber\": 4444,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 4444,
        \"host\": \"gp2-test-workspace-test-user.domain\",
        \"group\": \"gp2\",
        \"public\": true
      }
    ]
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 03 when changing the port Number 4']
SnapShot = """
[
  {
    \"Name\": \"proxy3333\",
    \"TargetPort\": 3333,
    \"LocalPort\": 99999
  }
]
"""

['auth-proxy controller ✅ success in normal context: 04 when changing the path  1']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1,
    \"network\": [
      {
        \"name\": \"nw1\",
        \"portNumber\": 1111,
        \"httpPath\": \"/abc\",
        \"targetPortNumber\": 99999,
        \"host\": \"gp1-test-workspace-test-user.domain\",
        \"group\": \"gp1\",
        \"public\": false
      },
      {
        \"name\": \"nw2\",
        \"portNumber\": 2222,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 2222,
        \"host\": \"gp2-test-workspace-test-user.domain\",
        \"group\": \"gp2\",
        \"public\": true
      }
    ]
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 04 when changing the path  2']
SnapShot = """
[
  {
    \"Name\": \"proxy1111\",
    \"TargetPort\": 1111,
    \"LocalPort\": 99999
  }
]
"""

['auth-proxy controller ✅ success in normal context: 04 when changing the path  3']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1,
    \"network\": [
      {
        \"name\": \"nw1\",
        \"portNumber\": 1111,
        \"httpPath\": \"/abc\",
        \"targetPortNumber\": 99999,
        \"host\": \"gp1-test-workspace-test-user.domain\",
        \"group\": \"gp1\",
        \"public\": false
      },
      {
        \"name\": \"nw2\",
        \"portNumber\": 2222,
        \"httpPath\": \"/cdf\",
        \"targetPortNumber\": 2222,
        \"host\": \"gp2-test-workspace-test-user.domain\",
        \"group\": \"gp2\",
        \"public\": true
      }
    ]
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 04 when changing the path  4']
SnapShot = """
[
  {
    \"Name\": \"proxy1111\",
    \"TargetPort\": 1111,
    \"LocalPort\": 99999
  }
]
"""

['auth-proxy controller ✅ success in normal context: 05 when changing the public flag 1']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1,
    \"network\": [
      {
        \"name\": \"nw1\",
        \"portNumber\": 1111,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 1111,
        \"host\": \"gp1-test-workspace-test-user.domain\",
        \"group\": \"gp1\",
        \"public\": true
      },
      {
        \"name\": \"nw2\",
        \"portNumber\": 2222,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 2222,
        \"host\": \"gp2-test-workspace-test-user.domain\",
        \"group\": \"gp2\",
        \"public\": true
      }
    ]
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 05 when changing the public flag 2']
SnapShot = """
[]
"""

['auth-proxy controller ✅ success in normal context: 05 when changing the public flag 3']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1,
    \"network\": [
      {
        \"name\": \"nw1\",
        \"portNumber\": 1111,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 1111,
        \"host\": \"gp1-test-workspace-test-user.domain\",
        \"group\": \"gp1\",
        \"public\": true
      },
      {
        \"name\": \"nw2\",
        \"portNumber\": 2222,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 99999,
        \"host\": \"gp2-test-workspace-test-user.domain\",
        \"group\": \"gp2\",
        \"public\": false
      }
    ]
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 05 when changing the public flag 4']
SnapShot = """
[
  {
    \"Name\": \"proxy2222\",
    \"TargetPort\": 2222,
    \"LocalPort\": 99999
  }
]
"""

['auth-proxy controller ✅ success in normal context: 06 wheh deleting the network rule 1']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1,
    \"network\": [
      {
        \"name\": \"nw1\",
        \"portNumber\": 1111,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 99999,
        \"host\": \"gp1-test-workspace-test-user.domain\",
        \"group\": \"gp1\",
        \"public\": false
      }
    ]
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 06 wheh deleting the network rule 2']
SnapShot = """
[
  {
    \"Name\": \"proxy1111\",
    \"TargetPort\": 1111,
    \"LocalPort\": 99999
  }
]
"""

['auth-proxy controller ✅ success in normal context: 06 wheh deleting the network rule 3']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 06 wheh deleting the network rule 4']
SnapShot = """
[]
"""

['auth-proxy controller ✅ success in normal context: 07 when deleting the workspace 1']
SnapShot = """
{
  \"Name\": \"test-workspace\",
  \"Namespace\": \"cosmo-user-test-user\",
  \"Spec\": {
    \"template\": {
      \"name\": \"template1\"
    },
    \"replicas\": 1,
    \"network\": [
      {
        \"name\": \"nw1\",
        \"portNumber\": 1111,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 99999,
        \"host\": \"gp1-test-workspace-test-user.domain\",
        \"group\": \"gp1\",
        \"public\": false
      },
      {
        \"name\": \"nw2\",
        \"portNumber\": 2222,
        \"httpPath\": \"/\",
        \"targetPortNumber\": 2222,
        \"host\": \"gp2-test-workspace-test-user.domain\",
        \"group\": \"gp2\",
        \"public\": true
      }
    ]
  },
  \"Status\": {
    \"instance\": {},
    \"phase\": \"Pending\",
    \"config\": {
      \"mainServicePortName\": \"main\",
      \"urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  }
}
"""

['auth-proxy controller ✅ success in normal context: 07 when deleting the workspace 2']
SnapShot = """
[
  {
    \"Name\": \"proxy1111\",
    \"TargetPort\": 1111,
    \"LocalPort\": 99999
  }
]
"""

['auth-proxy controller ✅ success in normal context: 07 when deleting the workspace 3']
SnapShot = """
{
  \"Name\": null,
  \"Namespace\": null,
  \"Spec\": null,
  \"Status\": null
}
"""

['auth-proxy controller ✅ success in normal context: 07 when deleting the workspace 4']
SnapShot = """
[]
"""
